common-pipeline-steps/src/main/scala/com/acxiom/pipeline/steps/MappingSteps.scala

```diff
@@ -0,0 +1,239 @@
+        exprList :+ expr(m.transform.get).as(m.outputField)
+      }
+    })
+
+    // return dataframe with all transforms applied
+    aliasedDF.select(finalExprs: _*)
+  }
+
+  /**
+    * converts data types from one dataframe to match data types of destination schema
+    * @param dataFrame          the input dataframe
+    * @param destinationSchema  the schema with the data types to adhere to
+    * @return   a new dataframe where all datatypes match destination schema data types
+    */
+  def convertDataTypesToDestination(dataFrame: DataFrame, destinationSchema: StructType): DataFrame = {
+    val columnExprs = dataFrame.schema.flatMap(a => {
+      val destination = destinationSchema.find(_.name == a.name)
+      if(destination.isEmpty || destination.get.dataType == a.dataType) {
+        // column doesn't exist in the destination
+        List(col(a.name))
+      } else {
+        // generate expression to select, cast, and rename
+        if(a.dataType != destination.get.dataType) {
+          logger.info(s"column '${a.name}' is being recast from ${a.dataType} to ${destination.get.dataType}")
+        }
+
+        List(col(a.name).cast(destination.get.dataType))
+      }
+    })
+
+    // return dataframe with attributes mapped to destination names and data types
+    dataFrame.select(columnExprs: _*)
+  }
+
+  /**
+    * adds placeholders for destination columns missing in dataframe
+    * @param dataFrame          the input dataframe
+    * @param destinationSchema  the destination schema with the desired columns
+    * @return   a new dataframe with placeholders
+    */
+  def addMissingDestinationAttributes(dataFrame: DataFrame, destinationSchema: StructType): DataFrame = {
+    // fold over destination schema starting with input dataframe columns
+    val finalExprs = destinationSchema.foldLeft(dataFrame.columns.map(x => col(x)))( (exprList, f) => {
+      if (dataFrame.columns.contains(f.name)) {
+        // already exists, no need to do anything to the expression list
+        exprList
+      } else {
+        // add null placeholder when destination field does not exist
+        logger.info(s"adding empty placeholder for column '${f.name}'")
+        exprList :+ lit("").cast(f.dataType).as(f.name)
+      }
+    })
+
+    // select expressions from dataframe and return
+    dataFrame.select(finalExprs: _*)
+  }
+
+  /**
+    * reorder columns in a dataframe to match order of a destination schema optionally adding new columns to the end
+    * @param dataFrame          the input dataframe to be reordered
+    * @param destinationSchema  the destination schema driving the column order
+    * @param addNewColumns      a flag determining whether missing attributes should be added
+    * @return   a new dataframe with column in order specified by destination schema
+    */
+  def orderAttributesToDestinationSchema(dataFrame: DataFrame, destinationSchema: StructType, addNewColumns: Boolean = true): DataFrame = {
```

Comment:
Should the code use the Schema class and then convert to a StructType for better flexibility?

Reply:
Probably, originally I tried to use StructType and not have Schema, but since global objects must be case classes, I had to create Schema so that it could be stored in the globals.  StructType in the non-published steps was carried over from the original.

The only ones that use StructType are the non-exposed methods, so it doesn't impact the users.  If we decide to expose those methods (see my response to your comment on that), I would argue that we might need both so that the user could pass a dataFrame.schema instead of creating a custom Schema object